src/IO/WriteBufferFromPocoSocketChunked.cpp

```diff
@@ -0,0 +1,207 @@
+        chunk_size_ptr = reinterpret_cast<decltype(chunk_size_ptr)>(pos);
+        pos += std::min(available(), sizeof(*chunk_size_ptr));
+
+        last_finish_chunk = chunk_size_ptr;
+
+        return;
+    }
+
+    /// Previously finished chunk wasn't sent yet
+    if (last_finish_chunk == chunk_size_ptr)
+    {
+        chunk_started = false;
+        LOG_TEST(log, "{} -> {} Chunk send ended.", ourAddress().toString(), peerAddress().toString());
+    }
+
+    /// Fill up current chunk size
+    setValue(chunk_size_ptr, toLittleEndian(static_cast<UInt32>(pos - reinterpret_cast<Position>(chunk_size_ptr) - sizeof(*chunk_size_ptr))));
+
+    if (!chunk_started)
+        LOG_TEST(log, "{} -> {} Chunk send started. Message {}, size {}",
+                ourAddress().toString(), peerAddress().toString(),
+                static_cast<unsigned int>(*(reinterpret_cast<char *>(chunk_size_ptr) + sizeof(*chunk_size_ptr))),
+                *chunk_size_ptr);
+    else
+    {
+        chunk_started = false;
+        LOG_TEST(log, "{} -> {} Chunk send continued. Size {}", ourAddress().toString(), peerAddress().toString(), *chunk_size_ptr);
+    }
+
+    LOG_TEST(log, "{} -> {} Chunk send ended.", ourAddress().toString(), peerAddress().toString());
+
+    if (available() < sizeof(*chunk_size_ptr))
+    {
+        finishing = available();
+        pos += available();
+        chunk_size_ptr = reinterpret_cast<decltype(chunk_size_ptr)>(pos);
+        return;
+    }
+
+    /// Buffer end-of-chunk
+    *reinterpret_cast<decltype(chunk_size_ptr)>(pos) = 0;
+    pos += sizeof(*chunk_size_ptr);
+    /// Initialize next chunk
+    chunk_size_ptr = reinterpret_cast<decltype(chunk_size_ptr)>(pos);
+    pos += std::min(available(), sizeof(*chunk_size_ptr));
+
+    last_finish_chunk = chunk_size_ptr;
+}
+
+WriteBufferFromPocoSocketChunked::~WriteBufferFromPocoSocketChunked()
+{
+    try
+    {
+        finalize();
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+
+void WriteBufferFromPocoSocketChunked::nextImpl()
+{
+    if (!chunked)
+    {
+        WriteBufferFromPocoSocket::nextImpl();
+        return;
+    }
+
+    /// next() after finishChunk ar the end of the buffer
+    if (finishing < sizeof(*chunk_size_ptr))
+    {
+        pos -= finishing;
+        /// Send current chunk
+        WriteBufferFromPocoSocket::nextImpl();
+        /// Send end-of-chunk directly
+        UInt32 s = 0;
+        socketSendBytes(reinterpret_cast<const char *>(&s), sizeof(s));
+
+        finishing = sizeof(*chunk_size_ptr);
+
+        /// Initialize next chunk
+        chunk_size_ptr = reinterpret_cast<decltype(chunk_size_ptr)>(working_buffer.begin());
+        nextimpl_working_buffer_offset = sizeof(*chunk_size_ptr);
+
+        last_finish_chunk = chunk_size_ptr;
+
+        return;
+    }
+
+    /// Send end-of-chunk buffered by finishChunk
+    if (offset() == 2 * sizeof(*chunk_size_ptr) && last_finish_chunk == chunk_size_ptr)
```

Comment:
Can you explain this case? If I understand correctly, it checks the situation when we sent a chunk in `next()`, then called `finishChunk()` that wrote end of chunk at the beginning of the buffer and then initialized new chunk, and then we called `next()` again. But how it differs from the situation `next() after finishChunk` on line 175? Seems like it will do the same logic except `last_finish_chunk = chunk_size_ptr;`

Reply:
> next generation of coders will be grateful for this :)

I would disagree, right now it seems to be totally unnecessary code for some special case that can be covered by more general use case later on line 175. Or I don't understand something

Reply:
Let's discuss it in DM, probably I miss something